File: core/src/rules/validate-url.ts
// core/src/rules/validate-url.ts
//
// Atomic rule: validates that a value is a parseable absolute URL.
//
// Uses Node's built-in `URL` constructor (zero deps). Throws → catch →
// INVALID_URL. We do NOT fetch the URL; we don't check whether it
// resolves. We only check structural parseability.
//
// Outcomes:
//   - non-string value         → INVALID_URL (error)
//   - string that won't parse  → INVALID_URL (error)
//   - parseable absolute URL   → no issue

import type { Issue } from "../types.js";
import { invalidUrl } from "../errors.js";

export function validateUrl(
  typeName: string,
  propertyName: string,
  value: unknown,
): Issue[] {
  if (typeof value !== "string") {
    return [invalidUrl(typeName, propertyName, value)];
  }
  try {
    // eslint-disable-next-line no-new
    new URL(value);
    return [];
  } catch {
    return [invalidUrl(typeName, propertyName, value)];
  }
}

File: python/src/schema_audit/rules/validate_url.py
"""Atomic rule: validate that a value is a parseable absolute URL.

Mirrors ``core/src/rules/validate-url.ts``. JS uses the built-in
``URL`` constructor; Python uses :func:`urllib.parse.urlparse` with the
same intent — structural parseability only, never a network call.

Outcomes:

- non-string value         → ``INVALID_URL`` (error)
- string that won't parse  → ``INVALID_URL`` (error)
- parseable absolute URL   → no issue
"""

from __future__ import annotations

from urllib.parse import urlparse

from ..errors import invalid_url
from ..types import Issue


def _is_absolute_url(value: str) -> bool:
    try:
        parsed = urlparse(value)
    except (ValueError, TypeError):
        return False
    # JS ``new URL("foo")`` throws if there's no scheme + host. Mirror that.
    return bool(parsed.scheme) and bool(parsed.netloc)


def validate_url(type_name: str, property_name: str, value: object) -> list[Issue]:
    if not isinstance(value, str):
        return [invalid_url(type_name, property_name, value)]
    if not _is_absolute_url(value):
        return [invalid_url(type_name, property_name, value)]
    return []
